File: algorithms/cpp/225.cpp
#include <iostream>
#include <queue>
using namespace std;

class Stack
{
private:
    queue<int> stack;
public:
    // Push element x onto stack.
    void push(int x)
    {
        stack.push(x);
        return;
    }

    // Removes the element on top of the stack.
    void pop()
    {
        queue<int> tmpData;
        while ( stack.size() > 1 )
        {
            tmpData.push(stack.front());
            stack.pop();
        }
        stack.pop();
        while ( !tmpData.empty() )
        {
            stack.push(tmpData.front());
            tmpData.pop();
        }
        return;
    }

    // Get the top element.
    int top()
    {
        queue<int> tmpData;
        while ( stack.size() > 1 )
        {
            tmpData.push(stack.front());
            stack.pop();
        }
        int returnVal = stack.front();
        tmpData.push(returnVal);
        stack.pop();
        while ( !tmpData.empty() )
        {
            stack.push(tmpData.front());
            tmpData.pop();
        }
        return returnVal;
    }

    // Return whether the stack is empty.
    bool empty()
    {
        return stack.empty();
    }
};

int main()
{
    return 0;
}

File: algorithms/rust/225.rs
use std::collections::VecDeque;

struct MyStack {
    swap: VecDeque<i32>,
    queue: VecDeque<i32>,
    top: i32,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
#[allow(dead_code)]
impl MyStack {
    fn new() -> Self {
        Self {
            swap: VecDeque::new(),
            queue: VecDeque::new(),
            top: 0,
        }
    }

    fn push(&mut self, x: i32) {
        self.top = x;
        self.queue.push_back(x);
    }

    fn pop(&mut self) -> i32 {
        let result = self.top;
        while let Some(value) = self.queue.pop_front() {
            if !self.queue.is_empty() {
                self.swap.push_back(value);
                self.top = value;
            }
        }
        std::mem::swap(&mut self.queue, &mut self.swap);
        result
    }

    fn top(&self) -> i32 {
        self.top
    }

    fn empty(&self) -> bool {
        self.queue.is_empty() && self.swap.is_empty()
    }
}

fn main() {
    println!("Hello, world!");
}
